src/types/checker/index.ts

```diff
@@ -0,0 +1,134 @@
+import { BadOperandTypesError, IncompatibleTypesError } from "../errors";
+import { Int, String, Type } from "../types";
+import { Node } from "../../ast/types/ast";
+
+export type Result = {
+  currentType: Type | null;
+  errors: Error[];
+};
+
+const OK_RESULT: Result = { currentType: null, errors: [] };
+
+const newResult = (
+  currentType: Type | null = null,
+  errors: Error[] = []
+): Result => ({ currentType, errors });
+
+export const check = (node: Node): Result => {
+  switch (node.kind) {
+    case "BinaryExpression": {
```

Comment:
Perhaps can consider creating a function for each case in the future so that the code will not be too nested and the whole switch statement will become more concise and easier to review.